fix: stop new rsvps cancelling attendance

ojeytonwilliams opened the pull request:
<!-- Please follow the below checklist and put an `x` in each of the boxes to agree with the statement, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [Chapter's contributing guidelines](https://github.com/freeCodeCamp/chapter/blob/main/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request targets the `main` branch of Chapter.

<!-- If your pull request closes a GitHub issue, replace the XXXXX below with the issue number. For e.g. Closes #12. The issue #12 will automatically get closed when this PR gets merged. -->

The issue was that we sent requests to the Google api with `attendees: [ { email: 'an@add.ress'} ]` rather than `attendees: [ { email: 'an@add.ress'}, responseStatus: ... ]`, that meant Google would remove their previous `responseStatus`.

The fix was to first get the event data, modify it and then send the updated data.

Tests to follow in another PR.

<!-- Tell us in detail about the changes you made and how it will affect the platform. -->

ojeytonwilliams committed:
feat: improve how Google service handles attendees

ojeytonwilliams committed:
refactor: less obnoxious function signatures

ojeytonwilliams committed:
refactor: verb form for function name

ojeytonwilliams committed:
fix: update event resolver for new Google service

ojeytonwilliams committed:
refactor: update calendar helper and use new api

ojeytonwilliams committed:
fix: update chapter user resolver for new service

ojeytonwilliams committed:
fix: stop sending emails in development

ojeytonwilliams committed:
refactor: param alias

ojeytonwilliams committed:
fix: use the correct function to create body

gikf commented on server/src/services/Google.ts:
```diff
@@ -189,98 +185,166 @@ function getStandardPatchBody({
-  });
+function filterFromAttendees(email: string) {
+  return (attendees: calendar_v3.Schema$EventAttendee[]) =>
+    attendees.filter((attendee) => attendee.email === email);
```
Are we leaving here other attendees, or just the one having `email`?
```suggestion
    attendees.filter((attendee) => attendee.email !== email);
```

ojeytonwilliams replied:
Oops. It's what you suggested.

Looks like I'm writing those tests now!

gikf reviewed (commented):
I like it very much in general.

ojeytonwilliams committed:
fix: remove attendee, not other attendees

Co-authored-by: Krzysztof G. <60067306+gikf@users.noreply.github.com>

ojeytonwilliams committed:
refactor: use more explicit name

ojeytonwilliams committed:
refactor: split Google into two modules

This is motivated primarily by testing (we can't mock createCalendarApi
otherwise), but it seems like a logical separation (setup the api in
one module and use it in another).

ojeytonwilliams committed:
fix: handle attendees correctly in parseEventData

ojeytonwilliams committed:
test: more tests for attendee manipulation

ojeytonwilliams committed:
test: check the event config is correct

ojeytonwilliams committed:
test: check createCalendar calls insert

ojeytonwilliams committed:
test: check createCalendarEvent is configured

Also clear all mocks between tests

ojeytonwilliams committed:
test: check createCalendarEvent's request data

ojeytonwilliams committed:
test: check updateCalendarEvent's request data

ojeytonwilliams committed:
test: check updateCalendarEvent ignores attendees

ojeytonwilliams committed:
refactor: remove use deleteCalendarEvent to cancel

As far as Google seems to be concerned, there isn't a difference between
setting the status to 'cancelled' and deleting via the delete endpoint.

ojeytonwilliams committed:
test: deleteCalendarEvent calls correct endpoint

ojeytonwilliams commented:
Phew! All done.

A few of the tests are a little trivial (does `deleteCalendarEvent` call `events.delete`? Yes. Yes it does), but overall I think they're reasonable.

I've not tested them locally, yet, I ran out of time. That's next on my list, but I figured it was worth sharing the code anyway.

gikf commented on server/tests/google.test.ts:
```diff
@@ -0,0 +1,286 @@
+      );
+    });
+
+    it("should do nothing if the email one of the attendee's", async () => {
```
```suggestion
    it("should do nothing if the email is not one of the attendee's", async () => {
```

ojeytonwilliams committed:
fix: typo in test description

Co-authored-by: Krzysztof G. <60067306+gikf@users.noreply.github.com>

Sboonny commented on server/src/util/calendar.ts:
```diff
@@ -2,64 +2,34 @@ import { inspect } from 'util';
 
-export const createCalendarEvent = async ({
+// TODO: consider pulling this back into the resolver (it's only used twice,
+// it's not that complex so I don't think we need to abstract it out)q
```
```suggestion
// it's not that complex so I don't think we need to abstract it out)
```

Sboonny reviewed (approved):
I think I couldn't test it, because `responseStatus` did fail for me, but I have gone to the calendar and activated google auth. 🤔

Well, doesn't matter anymore.  👍 Thank you @gikf, for reviewing this. 💚

ojeytonwilliams committed:
refactor: remove typo

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

ojeytonwilliams commented:
@Sboonny so we don't accidentally spam emails during Cypress testing, the attendees are set to `[]` when you're running the site in development mode.  Could it be that?

gikf reviewed (approved):
@ojeytonwilliams is there anything else you want to do here?

Sboonny commented:
Oh I think, I am the one causing the delay, because the talk about not being able to test it

ojeytonwilliams commented:
> is there anything else you want to do here?

Nope, I think it's fine.  I'll deploy it now and see if anything breaks.
